File: packages/domain/src/planproduction.ts
import { type ResolvedAlignment, pointAtStation } from "./alignment";
import { type Point } from "./geometry";
import { paperPerModel } from "./sheetview";
import { type Unit } from "./spatial";
import { type PaperUnit } from "./sheetsize";
import { type Sheet, type DrawingSet } from "./sheet";

/** A single rectangular View Frame showing a station range along an alignment. */
export interface ViewFrame {
  id: string;
  name: string;
  stationStart: number;
  stationEnd: number;
  center: Point;
  width: number;       // viewport width in model units
  height: number;      // viewport height in model units
  rotationDeg: number; // clockwise rotation angle in degrees to align viewport
}

/** A match line marking the page break boundary between two adjacent view frames. */
export interface PlanMatchLine {
  id: string;
  station: number;
  cutLine: [Point, Point]; // Left to Right normal cut line segment in model coordinates
  label: string;
}

/** A group organizing view frames and match lines along an alignment. */
export interface ViewFrameGroup {
  id: string;
  name: string;
  alignmentId: string;
  frames: ViewFrame[];
  matchLines: PlanMatchLine[];
}

/**
 * Automatically splits an alignment baseline into consecutive page segments (View Frames)
 * and calculates perpendicular boundary Match Lines.
 */
export function generateViewFrames(
  resolved: ResolvedAlignment,
  alignmentId: string,
  scaleId: string,
  viewportWidthIn: number,  // paper sheet viewport width in inches
  viewportHeightIn: number, // paper sheet viewport height in inches
  modelUnit: Unit,
  overlapRatio = 0.15,
): ViewFrameGroup {
  const paperUnit: PaperUnit = "in";
  // Model units per paper unit (inches)
  const scalePx = paperPerModel(scaleId, modelUnit, paperUnit);
  
  // Real world model dimensions of the viewport
  const modelWidth = scalePx > 0 ? viewportWidthIn / scalePx : 400;
  const modelHeight = scalePx > 0 ? viewportHeightIn / scalePx : 250;

  const stepDistance = modelWidth * (1 - overlapRatio);

  const frames: ViewFrame[] = [];
  const matchLines: PlanMatchLine[] = [];

  let station = resolved.startStation;
  let idx = 1;

  while (station < resolved.endStation) {
    const stationStart = station;
    const stationEnd = Math.min(resolved.endStation, station + modelWidth);
    const stationCenter = (stationStart + stationEnd) / 2;

    const centerPoint = pointAtStation(resolved, stationCenter);
    if (!centerPoint) break;

    // View rotation: set viewport horizontal to follow centerline bearing
    // heading is bearing, viewport horizontal aligns perpendicular to left normal
    const rotationDeg = centerPoint.bearing;

    frames.push({
      id: `frame-${alignmentId}-${idx}`,
      name: `View Frame #${idx} (STA ${Math.round(stationStart)} to ${Math.round(stationEnd)})`,
      stationStart,
      stationEnd,
      center: centerPoint.point,
      width: modelWidth,
      height: modelHeight,
      rotationDeg,
    });

    // Create match line at the end of the frame (if not the end of the alignment)
    if (stationEnd < resolved.endStation) {
      const edgePoint = pointAtStation(resolved, stationEnd);
      if (edgePoint) {
        const rad = (edgePoint.bearing * Math.PI) / 180;
        // Right normal vector
        const nx = Math.cos(rad);
        const ny = Math.sin(rad);

        // Match line extends perpendicular left/right of centerline
        const len = modelHeight / 2;
        const pLeft: Point = { x: edgePoint.point.x - len * nx, y: edgePoint.point.y - len * ny };
        const pRight: Point = { x: edgePoint.point.x + len * nx, y: edgePoint.point.y + len * ny };

        matchLines.push({
          id: `match-${alignmentId}-${idx}`,
          station: stationEnd,
          cutLine: [pLeft, pRight],
          label: `MATCH LINE STA ${Math.round(stationEnd)}`,
        });
      }
    }

    station += stepDistance;
    idx++;

    // Prevent infinite loop on edge cases
    if (stepDistance <= 0.01) break;
  }

  return {
    id: `vfg-${alignmentId}`,
    name: `View Frame Group - ${resolved.name}`,
    alignmentId,
    frames,
    matchLines,
  };
}

/** Creates a Sheet Set (DrawingSet) from a View Frame Group. */
export function createSheetSetFromFrames(
  group: ViewFrameGroup,
  drawingSetName: string,
): DrawingSet {
  const sheets: Sheet[] = group.frames.map((frame, i) => {
    const seq = i + 1;
    return {
      id: `sheet-${frame.id}`,
      number: {
        discipline: "C",
        type: 1, // 1-plans type digit in NCS
        sequence: seq,
      },
      title: `ROAD PLAN SHEET - PART ${seq}`,
      size: "arch-d", // Standard template layout size
      orientation: "landscape",
      scaleId: "1:500", // standard scale ID
      discipline: "C",
      viewportIds: [frame.id],
      revisions: [],
    };
  });

  return {
    id: `set-${group.id}`,
    name: drawingSetName,
    sheets,
    titleBlockDefaults: {
      projectName: "BLUEPRINT CORRIDOR PLAN",
      date: new Date().toLocaleDateString(),
      drawnBy: "Thoth AI",
    },
  };
}

File: crates/thoth-drawing/src/planproduction.rs
//! Plan production sheets — automatically splitting a linear alignment
//! (a road, a corridor) into consecutive **View Frames** (sheet-sized plan
//! windows) with **Match Lines** at the page breaks, and composing a
//! [`crate::sheet::DrawingSet`] from them.
//!
//! Port of `packages/domain/src/drawing/planproduction.ts`.
//!
//! ## Scope note
//!
//! [`create_sheet_set_from_frames`] is fully ported — it only needs
//! [`ViewFrameGroup`] (this module) and [`crate::sheet`] types. `generateViewFrames`
//! from the TS source is **not-yet-ported**: it needs `ResolvedAlignment` and
//! `pointAtStation` from `thoth-civil`, which this crate does not depend on.
//! See `STATUS.md`.

use thoth_spatial::Point;

use crate::sheet::{DrawingSet, Sheet, SheetNumber, TitleBlockDefaults};
use crate::sheetsize::Orientation;

/// A single rectangular View Frame showing a station range along an alignment.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct ViewFrame {
    pub id: String,
    pub name: String,
    pub station_start: f64,
    pub station_end: f64,
    pub center: Point,
    /// Viewport width in model units.
    pub width: f64,
    /// Viewport height in model units.
    pub height: f64,
    /// Clockwise rotation angle in degrees to align the viewport.
    pub rotation_deg: f64,
}

/// A match line marking the page break boundary between two adjacent view frames.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct PlanMatchLine {
    pub id: String,
    pub station: f64,
    /// Left-to-right normal cut line segment in model coordinates.
    pub cut_line: [Point; 2],
    pub label: String,
}

/// A group organizing view frames and match lines along an alignment.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct ViewFrameGroup {
    pub id: String,
    pub name: String,
    pub alignment_id: String,
    pub frames: Vec<ViewFrame>,
    pub match_lines: Vec<PlanMatchLine>,
}

/// Create a Sheet Set ([`DrawingSet`]) from a View Frame Group.
///
/// The TS original stamps `titleBlockDefaults.date` with
/// `new Date().toLocaleDateString()` at call time — an impure, host-locale-
/// dependent side effect with no fixed cross-platform format. This port
/// takes `date` as an explicit parameter instead: the caller supplies an
/// already-formatted date string, keeping this function pure and its output
/// deterministic and testable.
pub fn create_sheet_set_from_frames(
    group: &ViewFrameGroup,
    drawing_set_name: &str,
    date: &str,
) -> DrawingSet {
    use crate::drafting::DisciplineCode;

    let sheets: Vec<Sheet> = group
        .frames
        .iter()
        .enumerate()
        .map(|(i, frame)| {
            let seq = (i + 1) as u32;
            Sheet {
                id: format!("sheet-{}", frame.id),
                number: SheetNumber {
                    discipline: DisciplineCode::C,
                    r#type: 1,
                    sequence: seq,
                },
                title: format!("ROAD PLAN SHEET - PART {seq}"),
                size: "arch-d".to_string(),
                orientation: Orientation::Landscape,
                scale_id: "1:500".to_string(),
                discipline: DisciplineCode::C,
                viewport_ids: vec![frame.id.clone()],
                revisions: vec![],
                notes: vec![],
                keynote_ids: vec![],
            }
        })
        .collect();

    DrawingSet {
        id: format!("set-{}", group.id),
        name: drawing_set_name.to_string(),
        sheets,
        title_block_defaults: TitleBlockDefaults {
            project_name: "BLUEPRINT CORRIDOR PLAN".to_string(),
            client: None,
            location: None,
            drawn_by: Some("Thoth AI".to_string()),
            checked_by: None,
            date: date.to_string(),
            project_number: None,
            firm_lines: vec![],
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_group() -> ViewFrameGroup {
        ViewFrameGroup {
            id: "vfg1".to_string(),
            name: "VFG".to_string(),
            alignment_id: "align1".to_string(),
            frames: vec![
                ViewFrame {
                    id: "frame-align1-1".to_string(),
                    name: "View Frame #1".to_string(),
                    station_start: 0.0,
                    station_end: 400.0,
                    center: Point::new(200.0, 0.0),
                    width: 400.0,
                    height: 250.0,
                    rotation_deg: 0.0,
                },
                ViewFrame {
                    id: "frame-align1-2".to_string(),
                    name: "View Frame #2".to_string(),
                    station_start: 340.0,
                    station_end: 740.0,
                    center: Point::new(540.0, 0.0),
                    width: 400.0,
                    height: 250.0,
                    rotation_deg: 0.0,
                },
            ],
            match_lines: vec![],
        }
    }

    #[test]
    fn create_sheet_set_from_frames_builds_one_sheet_per_frame() {
        let set = create_sheet_set_from_frames(&sample_group(), "Corridor Plan", "2024-01-01");
        assert_eq!(set.sheets.len(), 2);
        assert_eq!(
            set.sheets[0].viewport_ids,
            vec!["frame-align1-1".to_string()]
        );
        assert_eq!(set.sheets[1].number.sequence, 2);
        assert_eq!(set.name, "Corridor Plan");
    }

    #[test]
    fn create_sheet_set_from_frames_of_empty_group_has_no_sheets() {
        let group = ViewFrameGroup {
            id: "vfg2".to_string(),
            name: "Empty".to_string(),
            alignment_id: "a".to_string(),
            frames: vec![],
            match_lines: vec![],
        };
        let set = create_sheet_set_from_frames(&group, "Empty Set", "2024-01-01");
        assert!(set.sheets.is_empty());
    }
}
